kolibri/auth/permissions.py

```diff
@@ -0,0 +1,349 @@
+        Note that the object may not yet have been saved to the database (as this may be a pre-save check)."""
+        return self._user_can_write_object(user, obj)
+
+    def user_can_read_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to read the provided <obj>."""
+        return False
+
+    def user_can_update_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to update the provided <obj>."""
+        return self._user_can_write_object(user, obj)
+
+    def user_can_delete_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to delete the provided <obj>."""
+        return self._user_can_write_object(user, obj)
+
+    def _user_can_write_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to write the provided <obj>.
+        Note that this is a private method, but can be overridden to modify the default behavior for
+        user_can_create_object, user_can_update_object, and user_can_delete_object in one fell swoop.
+        """
+        return False
+
+    def readable_by_user_filter(user, queryset):
+        """Applies a filter to the provided queryset, only returning items for which the user has read permission."""
+        raise NotImplementedError("Override `readable_by_user_filter` in your permission class before you use it.")
+
+
+class RoleBasedPermissions(BasePermissions):
+    """
+    Permissions class that defines a requesting user's permissions in terms of his or her role type with respect to
+    the object's associated User or Collection. These permissions are defined in terms of two data structures:
+    "permissions_by_role" (which lists which roles should have which permissions) and "target" (which
+    specifies how to find the associated User or Collection that is the target for checking the user's roles).
+    """
+
+    # For each permission type (read/write, or create/read/update/delete), specify a list of roles that should
+    # have that permission. By default, give permissions to no roles; override this or pass it into __init__
+    permissions_by_role = {
+        "read": [],
+        "write": [],
+    }
+
+    # Specify the "target" of the permissions, in relation to the object for which permissions are being granted.
+    # The "type" is either "user" or "collection", and the "field" is the field name (on the current object)
+    # through which the target (user or collection) will be referenced (or "." if the object itself is the target).
+    target = { # by default, the target is the user that is pointed to by the object's "user" field
+        "type": "user",
+        "field": "user",
+    }
+
+    def __init__(self, permissions_by_role=None, target=None):
+
+        if permissions_by_role:
+            self.permissions_by_role = permissions_by_role
```

Comment:
I'll have to look it up, but the instance property shadows the class property I assume?